docs: Add documentation on how to do EPT runs with our library. (#461)
foundation-model-stack/fms-hf-tuning: docs/ept.md

# Extended Pre Training Support

> Our library also supports Extended Pre-Training or (EPT) which is generally useful when users might want to to train a pretrained model on large number of samples. The training behaviour of EPT is similar to that of pretraining where users might wanna make sure the models runs through entire corpus of data available and be trained on whole set of tokens without any specific masking.

Reviewer: nit in `might want to to train`

`Our library also supports Extended Pre-Training (EPT), which is generally useful when users want to train a pretrained model on a large number of samples. `

Author: @Abhishek-TAMU fixed.

# Extended Pre Training Support
Our library also supports Extended Pre-Training or (EPT) which is generally useful when users might want to to train a pretrained model on large number of samples. The training behaviour of EPT is similar to that of pretraining where users might wanna make sure the models runs through entire corpus of data available and be trained on whole set of tokens without any specific masking.

## Packing support

We support training via `packing` dataset samples by specifing `--packing=True` in the command line parameters. Users can choose to specify `--max_seq_len=<value like 4k/8k>` to provide the maxium sequence length of each chunk post packing.

We provide below details on how to use different style of datasets with the library.

## Non-Tokenized Dataset

### Single Non-Tokenized Dataset

> Users can pass a single dataset to the library by using a [data_config](./advanced-data-preprocessing.md#data-config). While providing a `data_config` users need to provide `streaming=True` as the datasets for EPT are larger in size we request users to enable streaming datasets via this parameter to ease the load of not loading the entire dataset and overloading the memory.

Reviewer 2: ```suggestion
Users can pass a single dataset to the library by using a [data_config](./advanced-data-preprocessing.md#data-config). For the EPT scenario, users need to set `streaming=True` in the `datapreprocessor` section of their `data_config`. Since datasets for EPT are larger in size, they will often overload memory if the entire dataset is loaded all at once. The `streaming` parameter will mitigate this by loading the data in chunks. (See [data streaming](./advanced-data-preprocessing.md#data-streaming) documentation for more info)
```
Streaming docs don't exist yet (in pending PR) but once they do exist the link should work.

> Lets say you have a `JSONL` data file which contains text to be trained on in each line and you want to perform EPT on the same, you can create a `data_config` for the same in this manner,

Reviewer 2: ```suggestion
Lets say you have a `JSONL` data file which contains text to be trained on in each line that you want to perform EPT on, you can create a `data_config` for the dataset in this manner,
```

Example dataset,

```
{"Tweet":"@HMRCcustomers No this is my first job","ID":0,"Label":2,"text_label":"no complaint","output":"### Text: @HMRCcustomers No this is my first job\n\n### Label: no complaint"}
{"Tweet":"@KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.","ID":1,"Label":2,"text_label":"no complaint","output":"### Text: @KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.\n\n### Label: no complaint"}
...
```

Sample data config for the above use case.
```
dataprocessor:
    type: default
    streaming: true
datasets:
  - name: non_tokenized_text_dataset
    data_paths:
      - "<path-to-the-jsonl-dataset>"
```

And the commandline passed to the library should include following.

```
--data_config <path to the data config> --dataset_text_field "Tweet" --packing=True --max_seq_len 8192
```

Please note that for non tokenized dataset our code adds `EOS_TOKEN` to the lines, for e.g. `Tweet` column before passing that as a dataset.

### Multiple Non Tokenized Datasets

> If a user wants to utilise multiple datasets and want to [`sample`](./advanced-data-preprocessing.md#how-the-user-can-write-data-configs) the datasets. This can be acheived by specifying in the data config multiple datasets with differnt sampling ratios.

Reviewer 2: ```suggestion
If a user wants to utilize multiple datasets and want to [`sample`](./advanced-data-preprocessing.md#how-the-user-can-write-data-configs) the datasets. This can be achieved by specifying multiple datasets in the data config with different sampling ratios.
```

Sample data config for sampling among multiple datasets
```
dataprocessor:
    type: default
    streaming: true
    sampling_stopping_strategy: first_exhausted
    seed: 66
datasets:
  - name: non_tokenized_text_dataset_1
    sampling: 0.3
    data_paths:
      - "FILE_PATH"
  - name: non_tokenized_text_dataset_2
    sampling: 0.4
    data_paths:
      - "FILE_PATH"
  - name: non_tokenized_text_dataset_3
    sampling: 0.3
    data_paths:
      - "FILE_PATH"
```

> Please note we cover what different `sampling_strategies` mean and how to specify them in `data_config` as part of our document on [data mixing in advanced data preprocessing](./advanced-data-preprocessing.md#data-mixing)

Reviewer 2: ```suggestion
NOTE: More in-depth documentation of `sampling_stopping_strategy` and how to specify data mixing parameters in the `data_config` is covered in the [data mixing](./advanced-data-preprocessing.md#data-mixing) section of the advanced data preprocessing documentation
```

```
--data_config <path to the data config> --packing=True --max_seq_len 8192
```

### Pretokenised dataset with just input_ids

If your dataset contains just `input_ids` you can use one of our [data_handler](./advanced-data-preprocessing.md#preexisting-data-handlers) `duplicate_columns` to duplicate `input_ids` to `labels`.

In the same example above if the `pretokenised_dataset_1` did not contain `labels` column you can use this type of data config.

Sample dataconfig for this use case,

```
dataprocessor:
    type: default
    streaming: true
    sampling_stopping_strategy: first_exhausted
    seed: 66
datasets:
  - name: pretokenised_dataset_1
    sampling: 0.4
    data_paths:
      - "FILE_PATH"
    data_handlers:
      - name: duplicate_columns
        arguments:
          remove_columns: all
          batched: false
          fn_kwargs:
            old_column: "input_ids"
            new_column: "labels"
  - name: pretokenised_dataset_2
    sampling: 0.6
    data_paths:
      - "FILE_PATH"
```

The commandline for such use case would be similar to above.

```
--data_config <path to the data config> --packing=True --max_seq_len 8192

> ```

Reviewer 2: I would add a comment at the end including what version of fms-hf-tuning this document was created for and indicate that it might be outdated in future versions.

We provide below details on how to use different style of datasets with the library.

## Non-Tokenized Dataset

### Single Non-Tokenized Dataset
Users can pass a single dataset to the library by using a [data_config](./advanced-data-preprocessing.md#data-config). 
Lets say you have a `JSONL` data file which contains text to be trained on in each line that you want to perform EPT on, you can create a `data_config` for the dataset in this manner,

Example dataset,

```
{"Tweet":"@HMRCcustomers No this is my first job","ID":0,"Label":2,"text_label":"no complaint","output":"### Text: @HMRCcustomers No this is my first job\n\n### Label: no complaint"}
{"Tweet":"@KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.","ID":1,"Label":2,"text_label":"no complaint","output":"### Text: @KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.\n\n### Label: no complaint"}
...
```

Sample data config for the above use case.
```
dataprocessor:
    type: default
datasets:
  - name: non_tokenized_text_dataset
    data_paths:
      - "<path-to-the-jsonl-dataset>"

> ```

Reviewer 2: Should we include the appropriate data handler for the example dataset as well?

Author: ah yes..let me add that

> 

Reviewer 2: ```suggestion

NOTE: More in-depth documentation of the `streaming` parameter and the implications of setting it to `true` is covered in the [data streaming](./advanced-data-preprocessing.md#data-streaming) section of the advanced data preprocessing documentation

```
Streaming docs don't exist yet (in pending PR) but once they do exist the link should work.

Author: Streaming can be added later. Thanks @willmj I moved that commit
